design/pool-configuration.md

```diff
@@ -0,0 +1,386 @@
+
+From this point of view, we can enhance the allocation logic allowing users to reserve a given pool for a set of namespaces or for a set
+of services. This is more flexible and moves the focus on the cluster administrator, who is sets the MetalLB configuration
+in a multi tenant scenario.
+
+This, in combination with the other configuration options above, will provide a way to set specific options only to a namespace
+or to a service (i.e. specifying the communities for a given service).
+
+The issues related to this arguments:
+
+* [Issue: Pool affinity for Namespace](https://github.com/metallb/metallb/issues/383)
+* [Issue: Feature request: addresses assignment outside of Service definition](https://github.com/metallb/metallb/issues/616)
+* [Issue: allow selecting services to announce with the bgb speaker](https://github.com/metallb/metallb/issues/604)
+* [Issue: Dynamic BGP communities from services](https://github.com/metallb/metallb/issues/934)
+
+### Risks and Mitigations
+
+#### Conflicts with the current implementation
+
+The only current configuration option that will collide with what is being described here is the BGP peer node selector.
+
+Despite some overlap, it seems reasonable to maintain both filters and the only risk will be that if there is no intersection between filters,
+no risk of non-deterministic behavior seems to be present.
+
+#### Bad configuration
+
+##### Service allocation
+
+When a user specify both a pool to be allocated to a specific service and one with no selector, the service may receive an ip from the broader one.
+We need to implement an easy to understand and deterministic behaviour.
+
+##### Non overlapping filters
+
+Having an IP with no nodes advertising it is a risk. Some scenarios are:
+
+* Traffic Policy Local with pods not running on the same nodes selected by the selector
+* Node BGP peer selector with no intersection with the address selector
+
+It should be possible to detect non overlapping configurations, and raise an event when they happen. The metallb operator introduces a Validation WebHook
+that can be used to prevent those configurations to be created in first instance.
+
+### Conflicts with discovery
+
+The peer discovery feature currently being implemented in [this PR](https://github.com/metallb/metallb/pull/593) will introduce a way to
+fetch peers from the node's annotations.
+
+In order to be able to select them, the current proposed structure
+
+```yaml
+from-annotations:
+      - my-asn: example.com/local-asn
+        peer-asn: example.com/remote-asn
+        peer-address: example.com/address
+```
+
+will need to be extended with a name field (as much as the `peers` substructure) so it can be selected.
+
+### Conflicts with the FRR implementation
+
+All the BGP changes will happen from a "behavioural" point of view. Where needed, the changes will reflected in both the native and the new implementation.
```

Comment:
I think the FRR design proposal discussed not adding features to the native implementation when we started development on FRR. It might be that we need to do only a little to enable it but I wouldn't let the native implementation be a blocker.

Reply:
Agree, I just have the feeling that all the implementation will happen above the bgp interface.